# Challenges week 1

Now that you have some expereince with Jupyter notebooks and variables in Python, it's time for you to combine and apply this week's knowledge. You will start working on these challenges in the tutorial and will be asked to complete them by the end of the week. 

In each challange, you are asked to provide the programing solution to it as well as a technical interpretation explaining the steps taken and the result.


Some important notes for the challenges:

* If you get an error message, try to troubleshoot it (using Google often helps). If all else fails, go to the next excercise (but make sure to hand it in).
* We will make sure to provide feedback and improvement suggestions in the rubric. Make sure to read them; they will help you improve next week.

# Challenge 1

Below, you have a list that contains referals (places from which a visitor has entered a website) for one hour of the day. This is an example of digital trace data. Imagine that this is a list of referrals to the website of a news organization for one article. The list is ordered chronologically (i.e., first entries start at the full hour). 

Using the skills you have learned in the tutorial provide the following information:
* How many readers did the article have in one hour?
* Where did the first five visits come from?
* Where did the last 10 visitors come from?
* How many visitors came from the newsletter?

For each question, provide the necessary code and a short explanation of the choices you made.

In [1]:
visitors = ['homepage',
            'homepage',
            'homepage',
            'instagram app',
            'facebook app',
            'campaign utx=1389',
            'homepage',
            'facebook app',
            'homepage',
            'google search',
            'homepage',
            'facebook app',
            'homepage',
            'homepage',
            'homepage',
            'campaign utx=1389',
            'homepage',
            'homepage',
            'instagram app',
            'campaign utx=1389',
            'facebook.com',
            'facebook app',
            'google search',
            'homepage',
            'campaign utx=1389',
            'campaign utx=1389',
            'homepage',
            'instagram app',
            'homepage',
            'homepage',
            'homepage',
            'instagram app',
            'campaign utx=1389',
            'instagram app',
            'facebook app',
            'homepage',
            'facebook.com',
            'campaign utx=1389'
            'campaign utx=1389'
            'homepage',
            'homepage',
            'facebook.com',
            'instagram app',
            'instagram app',
            'campaign utx=1389',
            'homepage',
            'facebook.com',
            'newsletter',
            'newsletter',
            'homepage',
            'newsletter',
            'homepage',
            'newsletter',
            'facebook.com',
            'facebook.com',
            'homepage',
            'newsletter',
            'newsletter',
            'newsletter',
            'google search',
            'campaign utx=1389',
            'newsletter',
            'newsletter']

In [2]:
#Q1 - number of visitors
len(visitors)

61

By checking the lenght of the list using the `len` function, we can see how many visitors read the articles, namely 61.

In [3]:
#Q2 First 5 visitors
visitors[:5]

['homepage', 'homepage', 'homepage', 'instagram app', 'facebook app']

To find out the referals for the first five visits, we need to print the first 5 objects from the list. It looks like three visitors came from the homepage, one using the Instagram app and one using the Facebook app.

In [4]:
#Q3 last 10 visitors
visitors[-10:]

['facebook.com',
 'facebook.com',
 'homepage',
 'newsletter',
 'newsletter',
 'newsletter',
 'google search',
 'campaign utx=1389',
 'newsletter',
 'newsletter']

To find out the referals for the 10 last visits, we need to print the last 10 objects from the list. We co do it by specifying how many object we want to print counting from the end of the list (-10). It looks like two visitors came fromthe Facebook page, one from the homepage, five from the newsletter, one found the page via Google and one clicked on our campaign utx=1389. 

In [5]:
#Q4 number of visitors from the newsletter
visitors.count('newsletter')

9

We can use the method `.count` that allows us to count the number of occurences of an object in a list. We apply this method to the list, which allows us to see that 9 visitors came from the newsletter.

# Challenge 2

The table  contains information on grades for an assigment. 

| Name  | Grade |
|-------|-------|
| Emily | 6     |
| Ron   | 7     |
| John  | 7     |
| Amber | 8     |
| Emma  | 6     |
| David | 5     |
| Anne  | 6     |

1. Create a dictionary with the information from the table.
2. Using the dictionary you have created, write a code to compute average grade for the assignment. 
3. Rob was late with his assignment. Add his grade (7) to the dictionary.

For each question, provide the necessary code and a short explanation of the choices you made.

In [6]:
#Q1 Dictionary creation
grades = {'Emily': 6, 'Ron': 7, 'John': 7, 'Amber': 8, 'Emma':6, 'David': 5, 'Anne': 6}

To create a dictionary, I use `{}` and define the names as keys and the grades as values. I name the dictionary grades. 

In [7]:
#Q2 computaing average
sum(grades.values())/len(grades)

6.428571428571429

To compute average, I use functions seen in class. I need to divide the sum of all grades by the number of students. To access the gardes from the dictionary, I use `.values()`. This creates a list with all values from this dictionary. To get the sum of this list, I use the function `sum` that allows to create a sum of a list of integers. 

To get the number of students, I use `len` function on the dictionary. It returns how many key-value pairs are in the dictionary. 

Now, to get average, I need to devide the sum by the lenght of the dictionary. This shows me that the average grade was 6.43.

In [8]:
#Q3 adding key-value pair
grades['Rob'] = 7
grades

{'Emily': 6,
 'Ron': 7,
 'John': 7,
 'Amber': 8,
 'Emma': 6,
 'David': 5,
 'Anne': 6,
 'Rob': 7}

To add the grade of Rob, I need to add a key-value pair where `'Rob'` is the key and `7` the value. Printing the contents of the dictionary shows that the operation was succesful.

# Challenge 3

This is the most important challenge of this week, as we will use the concepts here to think how to organize our (digital) data. The paragraph below contains a lot of information. How would you organize this data in a way that you can visualize and work with it later? Consider using lists, dictionaries, tuples... or whatever makes sense to you!

Please explain why you decided to organize the data in the way that you did. In this explanation, consider both *technical* aspects (i.e., why lists, dictionaries, tuples or combinations thereof) and *substantive* aspects (i.e., why you structured the data in the way that you did).

```
Lake Aarens is 28 years old, works as an analyst at ING,
and is mother to a 3 year old boy. She is married to John.
On average, she spends 30 minutes a day on social media.
John Aarens is 32 years old, works as a journalist at NRC, 
and was born in the US. He has a 3 year old boy and is married to Lake. 
On average, he spends 45 minutes on social media.
Catharine Blakeslee is a student at the University of Amsterdam, born in 2000, 
she is not married and does not have children. She was born in the UK.
She does not use social media.
Walter Menno is 71 years old and is on pention. He is married to Anne
They do not have children. On average, Walter spends 1.5 hours a day on social media.
Anne Menno is 66 years old and works part-time as a teacher in a primary school.
She is married to Menno. On average, she spends 15 minutes a day on social media. 
```

In [9]:
people = {'Lake Aarens': {'Job': 'analyst', 'Age': 28, 'Country_birth': None, 'Kids': 1, 'marital_status': 1, 'SM_use':30},
          'John Aarens': {'Job': 'journalist', 'Age': 32, 'Country_birth': 'USA', 'Kids': 1, 'marital_status': 1, 'SM_use':45},
          'Catharine Blakeslee': {'Job': 'student', 'Age': 22, 'Country_birth': 'UK', 'Kids': 0, 'marital_status': 0, 'SM_use':0},          
          'Walter Menno': {'Job': 'pentionist', 'Age': 71, 'Country_birth': None, 'Kids': 0, 'marital_status': 1, 'SM_use':90}, 
          'Anne Menno': {'Job': 'teacher', 'Age': 66, 'Country_birth': None, 'Kids': 0, 'marital_status': 1, 'SM_use':15},           
}


One way to organize the data is to use nested dictionaries in which the names of the people are the keys and information about them are dictionaries with all variables we have available. In those dictionaries, the keys can be seen as "variable names" and values are the information about the people. This allows me to easily access necessary information
* For country of bith, the information is incomplete for some individuals. Therefore, their value is `None`. 
* For martial status, I assigned value 1 for married invididuals and 0 for not married.

In [10]:
people['Lake Aarens']['Age']

28

I can now access the data by refering to my dictionary people, then the key `'Lake Aarens'` and then the key `'Age'` within the dictionary associated with the key `'Lake Aarens'`.

*Note: this is only one way to organize this data. Multiple solutions to this challenge are possible.*